### ISSUE TITLE
salt base64 or raw bytes?

### ISSUE THREAD

**Issue by brocaar** [jpmens/mosquitto-auth-plug#261]
In one of my project I'm using a password string exactly as specified here:

```
PBKDF2$sha256$901$8ebTR72Pcmjl3cYq$SCVHHfqn9t6Ev9sE6RMTeF3pawvtGqTu
--^--- --^--- -^- ------^--------- -------------^------------------
  |      |     |        |                       |
  |      |     |        |                       +-- : hashed password
  |      |     |        +-------------------------- : salt
  |      |     +----------------------------------- : iterations
  |      +----------------------------------------- : hash function
  +------------------------------------------------ : marker
```

However I'm experiencing one issue. The `salt` at my side of the project is first base64 decoded and the "raw" bytes are used as input for the PBKDF2 function. However, it seems that the mosquitto-auth-plug is using the base64 encoded salt as input for `PBKDF2` function:

```c
	rc = PKCS5_PBKDF2_HMAC(password, strlen(password),
		(unsigned char *)salt, saltlen,
		iterations,
		evpmd, keylen, out);
	if (rc != 1) {
		goto out;
	}
```

Would it be an idea to make this configurable? I'm not sure if using base64 encoded salts as input for the PBKDF2 function is always be the case / is recommended. I've been looking at different examples and it seems that people use it in different ways.

Making this behaviour configurable would make this project usable for both implementations.

**Comment by brocaar** [comment]
^^ I've created a pull-request to make this configurable

**Comment by jpmens** [comment]
I think you're right: what a very good catch! (Crypto *is* hard; sighs).

**Comment by jpmens** [comment]
Closed via #262
